workflow/controller/operator.go

```diff
@@ -2118,6 +2135,25 @@ func (woc *wfOperationCtx) recordNodePhaseEvent(node *wfv1.NodeStatus) {
 	case wfv1.NodeSucceeded, wfv1.NodeRunning:
 		eventType = apiv1.EventTypeNormal
 	}
+	eventConfig := woc.controller.Config.NodeEvents
+	if eventConfig.SendAsPod != nil && *eventConfig.SendAsPod {
+		pod, err := woc.getPodByNode(node)
+		if err != nil {
+			woc.log.Infof("Error getting pod from workflow node: %s", err)
+		}
+		if pod != nil {
+			woc.eventRecorder.AnnotatedEventf(
+				pod,
+				map[string]string{
+					common.AnnotationKeyNodeType: string(node.Type),
+					common.AnnotationKeyNodeName: node.Name,
+				},
+				eventType,
+				fmt.Sprintf("WorkflowNode%s", node.Phase),
+				message,
+			)
+		}
+	}
 	woc.eventRecorder.AnnotatedEventf(
```

Comment:
This event is still sent even if the pod event is sent, resulting in sending two similar node events with the same `message` and `reason`, but a different `involvedObject`. As your unit tests show, this results in a somewhat noisier event stream:
```
"Normal WorkflowRunning Workflow Running",
"Normal WorkflowNodeRunning Running node no-dag-or-steps",
"Normal WorkflowNodeRunning Running node no-dag-or-steps",
"Normal WorkflowNodeSucceeded Succeeded node no-dag-or-steps",
"Normal WorkflowNodeSucceeded Succeeded node no-dag-or-steps",
"Normal WorkflowSucceeded Workflow completed",
```
In the case that a pod event is sent, I think it is better if the workflow event is not sent, but the Argo maintainers may have a different opinion. I assume this was a conscious design decision; was there any particular reason you chose to send both events?

Reply:
Yes, it was to make it backwards compatible. It felt better to accept more noise when opting in to this than to have particular events that were sending no longer sent.

Reply:
I don't think we should send the message twice. This can already produced a high-throughput of events, and this would 2x that throughput.

For pod nodes, just send it with the pod as involved object.

Reply:
> For pod nodes, just send it with the pod as involved object.

@alexec, does that mean this change would need to wait for a major Argo release? How does the backwards-incompatibility affect the release schedule?

Reply:
I'm not sure it really would be backwards incompatible. This is an opt-in feature, so that would be a change you acknowledge by opting in.

Reply:
Yes, that's a good point. It shouldn't change anything by default.